myhpom/models/cloudfactory.py

```diff
@@ -0,0 +1,146 @@
+                    else val
+                )
+                for key, val in self.__dict__.items()
+                if key[0] != '_'  # exclude utility vals like _state
+            }
+        )
+
+
+class CloudFactoryUnit(models.Model):
+    """Each CloudFactory run contains zero or more units."""
+
+    run = models.ForeignKey(
+        CloudFactoryRun,
+        on_delete=models.CASCADE,
+        help_text="The CloudFactory production run in which this unit is processed.",
+    )
+    status = models.CharField(
+        max_length=32, blank=True, default="", help_text="The status of the unit at CloudFactory."
+    )
+    created_at = models.DateTimeField(
+        blank=True, null=True, help_text="When the unit was created at CloudFactory."
+    )
+    processed_at = models.DateTimeField(
+        blank=True, null=True, help_text="When unit processing was finished at CloudFactory."
+    )
+    input = models.TextField(
+        blank=True,
+        default="",
+        help_text="JSON input to CloudFactory; use to validate the response.",
+    )
+    output = models.TextField(blank=True, default="", help_text="JSON output from CloudFactory")
+
+    def __repr__(self):
+        keys = [key for key in ['id', 'run_id'] if key in self.__dict__.keys()]
+        return u"%s(%s)" % (
+            self.__class__.__name__,
+            u", ".join([u"%s=%r" % (key, self.__dict__[key]) for key in keys]),
+        )
+
+    @property
+    def post_data(self):
+        """create data to POST the unit to CloudFactory"""
+        return self.data['input']
+
+    @property
+    def data(self):
+        """working with celery and email requires being able to provide json for the instance"""
+        return dict(
+            **{
+                key: (
+                    str(val)  # datetime.datetime and datetime.date as str
+                    if isinstance(val, datetime.datetime) or isinstance(val, datetime.date)
+                    else val
+                )
+                for key, val in self.__dict__.items()
+                if key[0] != '_'  # exclude utility vals like _state
+            }
+        )
+
+
+# The following is an unfortunate hack to overcome the lack of JSONField in Django 1.8:
+# convert the CloudFactoryUnit.input and .output fields to/from string for the database/python.
+# -- When we upgrade to Django 1.11 we can instead use django.contrib.postgres.fields.JSONField
+#   (but we'll need to be careful about migrating the database at that point!)
+
+
+def cloudfactory_unit_post_init_post_save(sender, instance, **kwargs):
+    """in python the input field is an object, not a string"""
+    if instance.input and (isinstance(instance.input, str) or isinstance(instance.input, unicode)):
+        instance.input = json.loads(instance.input)
+    if instance.output and (
+        isinstance(instance.output, str) or isinstance(instance.output, unicode)
+    ):
+        instance.output = json.loads(instance.output)
```

Comment:
Even if we were using django 1.11 these request/response fields should be strings -- b/c we may end up dealing with non json data (in particular on a response). Rather than using signals to modify the text fields directly here, provide explicit `json_input` and `json_output` properties on the models themselves.

Reply:
I don't think the CloudFactory API ever gives non-JSON responses, FWIW. But your approach would work.

Reply:
Karen's comments on Slack:

> re: catching ValueError from response.json()...it's probably desired in general for when the other side loses its mind and starts returning html or anything non-JSON. When we don't catch the exception and make an attempt to save the raw data we get very limited information on the failure (Exception Value: No JSON object could be decoded) and need to do some re-creation to see if the other side is saying anything useful at all. https://tracker.specialtyfood.com/issues/6303 is a recent case of this, where we had to spend time to investigate "well, what IS the other side saying?". As a result of having stuff like that happen I've pretty well moved towards all api usages trying to guard against this and save at least the beginning of any non-json data that we expected to be json...

> I've gotten in habit of writing wrappers like this for everywhere we use an API that's supposed to return json: https://github.com/nasft/nasft-services/blob/develop/experient/api_support.py#L24